CartService, OrderService 작성

yeonkyungJoo opened the pull request.

yeonkyungJoo committed:
CartService 작성

(cherry picked from commit cf7af10c7f005d6f134e906cb01d33c3a3c375eb)
(cherry picked from commit 3261bc838b83663084e01c38155ba323d49242e5)
(cherry picked from commit 364a946ad9deb47ecd7167a40906392107c31d96)

yeonkyungJoo committed:
OrderService 작성

(cherry picked from commit 311c76accb7f5c1e61decc061c00714e93440687)
(cherry picked from commit 1fd2d2ac32d409790cdef99236f06580affb76da)

yeonkyungJoo committed:
CartService, OrderService 완료

(cherry picked from commit 3c3225be38c42e501359882f0687dbd822588fb7)

f-lab-lyan commented:
어떻게 하면 잘 자를 수 있을지에 대해서 생각해 보기

CHANEE-personal commented on bbaemin-api/src/test/java/org/bbaemin/cart/service/CartServiceTest.java:
```diff
@@ -0,0 +1,121 @@
+                () -> assertThat(cart.getCartItemList().size()).isEqualTo(0)
+        );
+    }
+}
```
Ensure line feed at file end on Save 설정 해주시면 좋을거 같습니다.

CHANEE-personal commented on bbaemin-api/src/main/java/org/bbaemin/cart/repository/CartItemMemoryRepository.java:
```diff
@@ -0,0 +1,83 @@
+    private static final Map<Long, CartItem> map = new ConcurrentHashMap<>();
+    private static Long id = 0L;
+
+    static {
```
@PostConstruct 어노테이션도 활용해서 최초 데이터 생성할 수 있을 거 같습니다!

CHANEE-personal replied:
굳이 수정 안해주셔도 되고 참고만 해주시면 좋을 거 같습니다

f-lab-lyan replied:
`@PostContruct`는 별로 권장하는 패턴이 아니라고 생각하는데요. 왜 그럴까요? :-)

CHANEE-personal commented on bbaemin-api/src/main/java/org/bbaemin/cart/controller/CartController.java:
```diff
@@ -1,173 +1,73 @@
-                .orderAmount("12,390원")
-                .deliveryFee("3,000원")
-                .build();
+    @PutMapping("/{cartItemId}")
```
아래 plus, minus일 경우 mapping명을 다르게 설정을 해야할 거 같습니다! 
현재 ambiguous한거 같습니다

f-lab-lyan replied:
혹은 requestbody를 받아서 거기에 아이템이름과 수량을 적절히 넣을 수도 있을 것 같습니다.

yeonkyungJoo replied:
![image](https://user-images.githubusercontent.com/52401662/216803866-5075bf75-0bdd-4b30-86ff-b1766a3feda2.png)
@f-lab-lyan @CHANEE-personal 이렇게 변경하는 건 어떤가요?

CHANEE-personal commented on bbaemin-api/src/main/java/org/bbaemin/cart/service/CartService.java:
```diff
@@ -0,0 +1,74 @@
+        return new Cart(userId, cartItemList);
+    }
+
+    public CartItem addItem(Long userId, Long itemId) {
```
아직 고려할 단계는 아닌거 같지만 추후에 비회원이였을때 추가했었던 아이템도 회원가입을 했을때 동일하게 가져가는 기능까지 고려하면 좋을 거 같습니다.

f-lab-lyan replied:
비회원의 카트는 어떤 식으로 저장하는게 좋을까요?

CHANEE-personal commented on bbaemin-api/src/test/java/org/bbaemin/cart/service/CartServiceTest.java:
```diff
@@ -0,0 +1,121 @@
+
+class CartServiceTest {
+
+    CartItemRepository cartItemRepository = new CartItemMemoryRepository();
```
추후 testcase에서 아래 annotation 활용해서 주입받을 수 있을 거 같습니다! 참고만 부탁드립니다!
@TestConstructor
@RequiredArgsConstructor

f-lab-lyan commented on bbaemin-api/src/main/java/org/bbaemin/cart/controller/response/CartResponse.java:
```diff
@@ -1,40 +1,26 @@
-        this.orderAmount = orderAmount;
-        this.deliveryFee = deliveryFee;
+        this.orderAmount = cart.getFormattedOrderAmount();
+        this.deliveryFee = cart.getFormattedDeliveryFee(10000);
```
:)

f-lab-lyan commented on bbaemin-api/src/main/java/org/bbaemin/cart/repository/CartItemRepository.java:
```diff
@@ -0,0 +1,21 @@
+import java.util.List;
+import org.bbaemin.cart.vo.CartItem;
+
+public interface CartItemRepository {
```
#13 의 `UserRepository`와 같은 의견입니다.

f-lab-lyan commented on bbaemin-api/src/main/java/org/bbaemin/cart/service/CartService.java:
```diff
@@ -0,0 +1,74 @@
+        cartItemRepository.insert(cartItem);
+        return cartItem;
+    }
+/*
+    public Cart addItem(Long userId, Long itemId, int orderPrice) {
+//        Item item = itemRepository.findById(itemId);
+        CartItem cartItem = CartItem.builder()
+                .item(new Item_(itemId, "name", "description", 10000))
+                .userId(userId)
+                .orderCount(1)
+                .orderPrice(orderPrice)
+                .build();
+        cartItemRepository.insert(cartItem);
+        return getCart(userId);
+    }*/
```
주석 처리된 코드는 지워주시면 좋을 것 같습니다.

f-lab-lyan commented on bbaemin-api/src/main/java/org/bbaemin/order/vo/Item_.java:
```diff
@@ -0,0 +1,22 @@
+
+@ToString
+@Getter
+public class Item_ {
```
클래스 이름이 .... ??

yeonkyungJoo replied:
CartService, OrderService를 작성하면서 Item 클래스가 필요했는데
이 부분은 찬희님 담당이라 제가 일단 저렇게 작성을 했는데요.
@f-lab-lyan 이런 상황에서는 보통 어떻게 해결하는지 궁금합니다!

f-lab-lyan replied:
이 코멘트를 멘토링 할 때, 확인을 했네요. (뭔가 수를 써야할 것 같아요. 알람을 놓치네요 ㅜㅠ) 내 부분이 아닌 부분은 나중에 추가한다는 생각으로 TODO를 붙이고 넘어갑니다. 그리고 나중에 Item이 추가되면, 그 때, 추가하는 방식으로 작업하는 것 같네요.

f-lab-lyan commented on bbaemin-api/src/main/java/org/bbaemin/cart/vo/Cart.java:
```diff
@@ -0,0 +1,36 @@
+    }
+
+    private int getDeliveryFee(int criteria) {
+        return getOrderAmount() >= criteria ? 0 : 3000;
```
두 가지 생각해볼 문제가 있는데요.
1. 만약 3000원이라는 값이 변경되면, 어떻게 할 것인가? 혹은 저 값을 동적으로 생성해 줘야하는 상황이라면 어떻게 해야할까?
2. Cart라는 VO에서 DeliveryFee를 결정하는 것이 괜찮을까?

yeonkyungJoo replied:
DeliveryFeeService를 생성해서 DeliveryFee를 가져오는 메소드를 새로 추가했습니다.

f-lab-lyan commented on bbaemin-api/src/test/java/org/bbaemin/order/service/OrderServiceTest.java:
```diff
@@ -0,0 +1,132 @@
+    }
+
+    @Test
+    void cancelOrder() {
+        // given
+        CartItem cartItem = cartService.addItem(1L, 1L);
+        Order order = orderService.order(1L, CreateOrderRequest.builder()
+                .deliveryAddress("서울시 강동구")
+                .phoneNumber("010-1111-2222")
+                .email("user@email.com")
+                .messageToRider("감사합니다")
+                .paymentMethod("신용/체크카드")
+                .build());
+        // when
+        orderService.cancelOrder(1L, order.getOrderId());
+        // then
+        Order canceled = orderRepository.findById(order.getOrderId());
+        assertThat(canceled.getStatus()).isEqualTo(OrderStatus.CANCEL_ORDER);
+    }
```
다른 테스트도 비슷한데, 테스트를 하고자 하는 범위 혹은 정확한 기능이 무엇인지 좀 모호해 보입니다. 체크하는 부분이 Repository에 데이터 상태가 정확한지를 테스트 하는데, 이렇게 되면, 서비스 테스트를 위해서 Repository가 필요하게 되고, 그럼 UnitTest라기 보다는 IntegrationTest에 더 가까워 지게 되는데요. 어떻게 하면, Service만 테스트 할 수 있을지 생각해 보면 좋을 것 같습니다.

@yeonkyungJoo @CHANEE-personal 이 부분은 멘토링 시간에 얘기를 해보면 좋을 것 같습니다.

f-lab-lyan commented on bbaemin-api/src/main/java/org/bbaemin/cart/service/CartService.java:
```diff
@@ -0,0 +1,74 @@
+    public CartItem addItem(Long userId, Long itemId) {
+//        Item item = itemRepository.findById(itemId);
+        CartItem cartItem = CartItem.builder()
+                .item(new Item_(itemId, "name", "description", 10000))
```
이렇게 item을 CarItem에 저장하면, CartItem에 저장되어 있는 item의 실제 정보가 바뀌는 경우에는 어떻게 반영해야할까요?

yeonkyungJoo replied:
실제 정보가 바뀌는 부분은 로직에서 itemId로 새로 조회를 해야 합니다. 해당 부분은 Cart에 넣었던 당시의 Item 정보를 그대로 넣어주기 위한, 의도된 부분입니다.

yeonkyungJoo committed:
Cart API 수정

yeonkyungJoo committed:
request/responseDTO Service에서 제거

yeonkyungJoo committed:
Entity에서 비즈니스 로직 제거

yeonkyungJoo committed:
Repository 인터페이스 제거

yeonkyungJoo committed:
StringUtils 생성

yeonkyungJoo committed:
DeliveryFeeService 추가로 Cart에서 DeliveryFee 분리

yeonkyungJoo committed:
OrderServiceTest - 단위테스트 재작성

yeonkyungJoo committed:
DeliveryFeeService - deliveryFee를 cartItemList로 조회하는 메소드로 변경

yeonkyungJoo committed:
1. Cart VO 제거
2. CartService를 CartItemService로 rename
3. CartResponse의 생성자 파라미터를 cart가 아닌 cartItemList로 변경

yeonkyungJoo committed:
1. CartItem에서 비즈니스 로직 제거
2. CartItem에서 Item 필드 제거 / itemName, itemDescription 필드 추가

yeonkyungJoo committed:
1. Order에서 비즈니스 로직 제거
2. Order에서 orderItemList 필드 제거 → orderItemRepository에서 orderId로 조회하는 것으로 변경
3. Order에서 discountCouponIdList 분리
4. OrderResponse 생성자에 orderItemList 추가

yeonkyungJoo committed:
1. OrderItem에서 Item 필드 제거
2. OrderItem에 주문 당시의 Item 정보 저장을 위해 itemName, itemDescription 필드 추가
3. OrderItem에서 비즈니스 로직 제거

f-lab-lyan commented on bbaemin-api/src/main/java/org/bbaemin/cart/controller/CartController.java:
```diff
@@ -1,173 +1,72 @@
-                .deliveryFee("3,000원")
-                .build();
+    @PostMapping("/items")
+    public CartResponse addItem(@PathVariable Long userId, @RequestBody CreateCartItemRequest createCartItemRequest) {
```
`@PathVariable Long userId`?

f-lab-lyan commented on bbaemin-api/src/main/java/org/bbaemin/cart/controller/CartController.java:
```diff
@@ -1,173 +1,72 @@
+import java.util.List;
 
-@RequestMapping("/api/v1/cart")
+@RequestMapping("/api/v1/users/{userId}/cart")
```
아! 여기 있었군요. 음, 저는 개인적으로 좋아하는 방식은 아닙니디만, @CHANEE-personal 님은 어떠세요?

f-lab-lyan commented on bbaemin-api/src/main/java/org/bbaemin/util/StringUtils.java:
```diff
@@ -0,0 +1,18 @@
+import java.time.LocalDateTime;
+import java.time.format.DateTimeFormatter;
+
+public class StringUtils {
```
코드 자체는 문제는 없어보이는데요. 곧이 public static으로 만들 이유가 있었을까요? 그냥 적당히 Service로 만드는게 더 통일성이 있지 않았을까요? Util로 만들어서 static 메소스들을 넣기 시작하면, 지금 작성한 것처럼 사실은 상관없는 두 개의 로직이 같이 들어가는 문제가 생길 것 같은데요. 이 부분에 대해서는 어떻게 생각하세요?

f-lab-lyan commented on bbaemin-api/src/main/java/org/bbaemin/cart/controller/CartController.java:
```diff
@@ -1,173 +1,72 @@
-//                .orderItemList(Collections.emptyList())
-                .build();
+    public CartResponse clear(@PathVariable Long userId) {
+        cartItemService.clear(userId);
+        return getCart(userId);
```
위에도 비슷한 로직이 있는데, 그냥 clear에서 CartResponse를 리턴하게 하는게 어떨까요? 물론 CartResponse를 Service에서 만드는게 맞는지는 또 다른 문제이긴 합니다.

f-lab-lyan commented on bbaemin-api/src/test/java/org/bbaemin/cart/service/CartItemServiceTest.java:
```diff
@@ -0,0 +1,14 @@
+import org.mockito.InjectMocks;
+import org.mockito.Mock;
+
+class CartItemServiceTest {
```
테스트는 이제 만들 것이라는 것인가요? :-)

yeonkyungJoo replied:
테스트 추가했습니다!

f-lab-lyan commented on bbaemin-api/src/test/java/org/bbaemin/order/service/OrderServiceIntegrationTest.java:
```diff
@@ -0,0 +1,164 @@
+import static org.bbaemin.order.enums.OrderStatus.COMPLETE_ORDER;
+import static org.junit.jupiter.api.Assertions.assertAll;
+
+class OrderServiceIntegrationTest {
```
이 테스트는 아래에 있는 테스트랑 내용은 거의 똑같은 느낌인데, 특별히 두 벌의 테스트를 넣으신 이유가 있나요? 나중에 Repository가 실제로 구현되었을 때도 테스트를 할꺼라던가 ...

f-lab-lyan commented on bbaemin-api/src/main/java/org/bbaemin/order/service/OrderService.java:
```diff
@@ -0,0 +1,85 @@
+        order.setOrderAmount(orderAmount);
+        order.setDeliveryFee(deliveryFee);
+
+        if (cartItemList.size() > 1) {
+            order.setDescription(String.format("%s 외 %d개", cartItemList.get(0).getItemName(), cartItemList.size() - 1));
+        } else {
+            order.setDescription(cartItemList.get(0).getItemName());
+        }
```
이 부분은 백엔드에서 하기에는 너무 자세한 느낌인데, 연경님의 개인적인 경험이 궁금하네요. 도대체 무슨 일이 있으셨길래 이런 로직을 넣으신걸까요? :-)

f-lab-lyan commented on bbaemin-api/src/main/java/org/bbaemin/cart/controller/response/CartItemResponse.java:
```diff
@@ -0,0 +1,35 @@
+    private int totalOrderPrice;
+    private String totalOrderPriceStr;
+
+    public int getTotalOrderPrice() {
+        return this.totalOrderPrice;
+    }
+
+    public CartItemResponse(CartItem cartItem) {
+        this.itemName = cartItem.getItemName();
+        this.itemDescription = cartItem.getItemDescription();
+        this.orderPrice = cartItem.getOrderPrice();
+        this.orderPriceStr = getFormattedPrice(cartItem.getOrderPrice());
+        this.orderCount = cartItem.getOrderCount();
+        this.totalOrderPrice = cartItem.getOrderPrice() * cartItem.getOrderCount();
+        this.totalOrderPriceStr = getFormattedPrice(cartItem.getOrderPrice() * cartItem.getOrderCount());
+    }
```
nit: 다른 코드들은 Constructor가 제일 위해 있는 것 같아요. :-)

f-lab-lyan commented on bbaemin-api/src/main/java/org/bbaemin/cart/service/CartItemService.java:
```diff
@@ -0,0 +1,55 @@
+        return cartItemRepository.findByUserId(userId);
+    }
+
+    public CartItem addItem(Long userId, Long itemId) {
```
만약 똑같은 아이템을 두 번 넣으면 어떻게 되나요? 테스트에 추가해보시면 좋을 것 같네요.

yeonkyungJoo replied:
![image](https://user-images.githubusercontent.com/52401662/219855375-0fc95646-ddef-4c54-bfbc-bef3090ffd3b.png)
예외 처리에 관한 부분일까요?

f-lab-lyan replied:
예외처리라기 보다는 비지니스 로직에 관련된 얘기 입니다. 실제로 사용자가 사이트를 돌다가 똑같은 아이템을 두 번 넣을 수도 있지 않을까 싶어서요.

f-lab-lyan commented on bbaemin-api/src/main/java/org/bbaemin/cart/service/DeliveryFeeService.java:
```diff
@@ -0,0 +1,25 @@
+    private static final int CRITERIA = 10000;
+    private static final int BASIC_COST = 3000;
+
+    public int getDeliveryFee(List<CartItem> cartItemList) {
+        int orderAmount = cartItemList.stream()
+                .mapToInt(cartItem -> cartItem.getOrderPrice() * cartItem.getOrderCount()).sum();
+        return getDeliveryFee(orderAmount);
+    }
```
이 건 사용되는 코드인가요? 적어도 이 PR에서는 안 보이는 것 같은데요. (제가 못 찾은거라면 알려주세요 ㅜㅠ)

yeonkyungJoo replied:
![image](https://user-images.githubusercontent.com/52401662/219850886-dcad8912-72e2-4f2f-af57-0529fda7a9fd.png)
장바구니를 조회할 때 담은 상품을 토대로 계산해서 CartResponse로 리턴하려고 했는데요!
혹시 이것도 front 쪽에서 하는 게 더 나을까요?

f-lab-lyan replied:
오 .. 제가 못 찾은 거였군요. 아뇨. 이 부분은 백엔드에서 해주는게 더 좋을 것 같아요. 왜냐하면, 상품에 따라서 배송비가 달라지거나 하는 등의 비지니스를 프론트에서 가져가면, 쉽지 않을 것 같다는 생각이 드네요.

f-lab-lyan commented on bbaemin-api/src/main/java/org/bbaemin/order/controller/response/OrderSummaryResponse.java:
```diff
@@ -0,0 +1,28 @@
+        this.orderId = order.getOrderId();
+        this.status = order.getStatus().getName();
+        this.description = order.getDescription();
+
```
nit: Remove the blank line.

f-lab-lyan commented on bbaemin-api/src/main/java/org/bbaemin/order/repository/OrderItemRepository.java:
```diff
@@ -0,0 +1,54 @@
+@RequiredArgsConstructor
+public class OrderItemRepository {
+
+    private static final Map<Long, OrderItem> map = new ConcurrentHashMap<>();
+    private static Long id = 0L;
```
문득 생각난 것인데, 왜 이 부분을 `private static final`로 만든건가요? 그냥 `private final`이면 충분하지 않았을까요? `@Component`니까?

yeonkyungJoo replied:
습관적으로 static을 쓴 것 같습니다^^
코드 작성할 때 더 신경 쓰도록 하겠습니다!

yeonkyungJoo replied:
그럼 아래 id도 static이 아니여도 되는 거 맞나요?

f-lab-lyan replied:
네, 그럴 것 같습니다. :-)

f-lab-lyan commented on bbaemin-api/src/main/java/org/bbaemin/order/repository/OrderItemRepository.java:
```diff
@@ -0,0 +1,54 @@
+    }
+
+    public void deleteByOrderId(Long orderId) {
+        Iterator<Map.Entry<Long, OrderItem>> iterator = map.entrySet().stream().iterator();
+        while (iterator.hasNext()) {
+            Map.Entry<Long, OrderItem> next = iterator.next();
+            if (next.getValue().getOrderId().equals(orderId)) {
+                map.remove(next.getKey());
+            }
+        }
```
이와 stream을 쓴 김에 filter를 써보면 어떨까 싶네요. :-)

yeonkyungJoo replied:
`map.entrySet().removeIf(entry -> entry.getValue().getOrderId().equals(orderId));`
수정했습니다!

f-lab-lyan reviewed (changes requested):
몇 가지 생각해 볼 지점들이 있어서 변경 요청 드렸습니다. 코멘트 확인해 주세요.

yeonkyungJoo committed:
Constructor 위치 변경

yeonkyungJoo committed:
response API 수정

yeonkyungJoo committed:
deleteByOrderId - removeIf를 사용해서 처리

yeonkyungJoo committed:
@Component 클래스의 필드 static 제거

yeonkyungJoo committed:
테스트코드 작성

f-lab-lyan reviewed (approved):
나머지 변경 사항들은 새로운 PR에서 하는 걸로하죠 :-) 고생하셨습니다.

f-lab-lyan commented:
다음 작업을 위해서 제가 머지하겠습니다. @CHANEE-personal
